### ISSUE TITLE
[FEATURE] Replace BPR full-catalogue scoring with sampled-softmax or item-only embedding lookup

### ISSUE THREAD

**Issue by sachncs** [sachncs/morel#23]
## Is your feature request related to a problem?

`Recommendation.step` (and `.validate`) score the user against the full catalogue on every step:

```python
# morel/train/recommendation.py:99-101
scores = self.model(users, torch.arange(self.items, device=self.device), self.ui)
pos_scores = scores[torch.arange(users.shape[0], device=self.device), pos]
neg_scores = scores[torch.arange(users.shape[0], device=self.device), neg]
```

A full forward pass produces a `(B_u, items)` matrix; only two columns (`pos` and `neg`) are used. For a catalogue of 100k items and batch 1024, that's a 400 MB score tensor per step. At item scale this becomes the dominant memory and time cost of recommendation training.

## Describe the Solution You'd Like

Either:

- Sample N negatives per positive inside the trainer (sampled-softmax BPR), or
- Score only the `pos` and `neg` indices by indexing the item embedding `self.model.item_emb.weight[torch.tensor([pos, neg])]` and doing the dot-product explicitly.

Both avoid the `items`-sized score matrix. Sampled-softmax additionally gives a better gradient signal.

## Describe Alternatives You've Considered

- Keep the full score and rely on hardware (out of scope for a research library).
- Subsample items per epoch (statistical bias).

## Additional Context

Engineering audit finding `PF-04` (High). The LightGCN model already has the per-user / per-item embeddings we need; the trainer just needs to index them directly.

## Checklist

- [x] I have searched the existing issues.
- [x] I have provided a clear description of the problem and proposed solution.
- [x] I have considered alternative approaches.

**Comment by sachncs** [comment]
Resolved in master. The relevant fix is split across one or two commits; see the change history since the issue was opened.
